core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -264,16 +239,95 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
     authorizerLogger.debug(s"Principal = $principal is $permissionType Operation = $operation from host = $host on resource = $resource")
   }
 
+  /**
+    * Safely updates the resources acls by ensureing reads and writes respect the expected zookeeper version.
+    * Continues to retry until it succesfully updates.
+    *
+    * @param resource the resource to change acls for
+    * @param getNewAcls function to transform existing acls to new acls
+    * @return boolean indicating if a change was made
+    */
+  private def updateResourceAcls(resource: Resource)(getNewAcls: Set[Acl] => Set[Acl]): Boolean = {
+    val path = toResourcePath(resource)
+
+    var currentVersionedAcls =
+      if(aclCache.contains(resource))
+        getAclsFromCache(resource)
+      else
+      getAclsFromZk(resource)
+    var newVersionedAcls = currentVersionedAcls
+    var writeComplete = false
+    while (!writeComplete) {
+      val newAcls = getNewAcls(currentVersionedAcls.acls)
+      val data = Json.encode(Acl.toJsonCompatibleMap(newAcls))
+      try {
+        val (updateSucceeded, updateVersion) = zkUtils.conditionalUpdatePersistentPathIfExists(path, data, currentVersionedAcls.version)
+        if(!updateSucceeded) {
+          trace(s"Failed to update acls for $resource. Used version ${currentVersionedAcls.version}. Reading data and retrying update.")
+          currentVersionedAcls = getAclsFromZk(resource);
+        } else {
+          if(newAcls.isEmpty) {
+            trace(s"Deleting path for $resource because it had no acls remaining")
+            zkUtils.deletePath(path) // It would be safest to use the versioned request, but this zkClient doesn't support it
```

Comment:
I agree that you do want to use the version here. The options I see are that we get the handle directly through ZkUtils.zkConnection.getZookeeper and call delete using the raw zk handle or that we patch ZkClient. I had a look at ZkClient and it should be straightforward to do it. Perhaps we can do the former until we have a new version of ZkClient with the versioned delete call and propose the change to ZkClient in parallel. I can help with that if you like the direction.